packages/lesswrong/lib/collections/users/helpers.ts

```diff
@@ -306,47 +307,84 @@ export const userGetLocation = (currentUser: UsersCurrent|DbUser|null): {
-// is used first. If the user is not logged in, then no location is available
-// for server-side rendering, but we can try to get a location client-side
-// using the browser geolocation API. (This won't necessarily work, since not
-// all browsers and devices support it, and it requires user permission.)
-export const useUserLocation = (currentUser: UsersCurrent|DbUser|null) => {
+/**
+ * Return the current user's location, by checking a few places.
+ *
+ * If the user is logged in, the location specified in their account settings is used first.
+ * If the user is not logged in, then no location is available for server-side rendering,
+ * but we can check if we've already saved a location in their browser's local storage.
+ *
+ * If we've failed to get a location for the user, finally try to get a location
+ * client-side using the browser geolocation API.
+ * (This won't necessarily work, since not all browsers and devices support it, and it requires user permission.)
+ * This step is skipped if the "dontAsk" flag is set, to be less disruptive to the user
+ * (for example, on the forum homepage).
+ *
+ * @param {UsersCurrent|DbUser|null} currentUser - The user we are checking.
+ * @param {boolean} dontAsk - Flag that prevents us from asking the user for their browser's location.
+ *
+ * @returns {Object} location
+ * @returns {number} location.lat - The user's latitude.
+ * @returns {number} location.lng - The user's longitude.
+ * @returns {boolean} location.loading - Indicates that we might have a known location later.
+ * @returns {boolean} location.known - If false, then we're returning the default location instead of the user's location.
+ * @returns {string} location.label - The string description of the location (ex: Cambridge, MA, USA).
+ * @returns {Function} location.setLocation - Function to set the location directly.
+ */
+export const useUserLocation = (currentUser: UsersCurrent|DbUser|null, dontAsk?: boolean): {
+  lat: number,
+  lng: number,
+  loading: boolean,
+  known: boolean,
+  label: string,
+  setLocation: Function
+} => {
+  // default is Berkeley, CA
   const placeholderLat = 37.871853
   const placeholderLng = -122.258423
+  const defaultLocation = {lat: placeholderLat, lng: placeholderLng, loading: false, known: false, label: null}
   
   const currentUserLat = currentUser && currentUser.mongoLocation && currentUser.mongoLocation.coordinates[1]
   const currentUserLng = currentUser && currentUser.mongoLocation && currentUser.mongoLocation.coordinates[0]
 
   const [location, setLocation] = useState(() => {
-    if (currentUserLat && currentUserLng) {
+    if (currentUserLat && currentUserLng && currentUser?.location) {
       // First return a location from the user profile, if set
-      return {lat: currentUserLat, lng: currentUserLng, loading: false, known: true}
+      return {lat: currentUserLat, lng: currentUserLng, loading: false, known: true, label: currentUser.location}
```

Comment:
I'm surprised at this addition.

Reply:
Can you expand on this? In what way are you surprised, and what are you surprised about?

Reply:
I'm not sure what I was saying. No points for clarity here. I think maybe it was the location object vs location string thing.